rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,241 @@
+  | ObjectShorthandProps<JSX.IntrinsicElements['span'], HTMLSpanElement, /*as*/ 'span'>
+  | ObjectShorthandProps<JSX.IntrinsicElements['a'], HTMLAnchorElement, /*as*/ 'a'>;
+```
+
+`ARIAButtonProps` is a union between types from `button`, `div`, `span` and `anchor` native elements.
+
+> ⚠️ using union types for declaring properties is not perfect https://catchts.com/unions
+
+### No reason for `getNativeElementProps`
+
+The problem reported on [Widening Types](#⚠️-Widening-types-problem-⚠️), makes this a little bit controversial.
+
+Although types should be enough to ensure that only supported properties will be passed to a _slot_, [Widening Types problem](#⚠️-Widening-types-problem-⚠️) make it not so much true for `root` at least. Besides `root` is safe to say that `getNativeelementProps` is not necessary.
+
+### Add references to all _slots_
+
+Right now _ref_ is not supported on _Typings_ for _slots_. This proposes that _ref_ should be part of the internal interface of `ObjectShorthandProps` to allow access of references to internal slots of a given component.
+
+```ts
+type ObjectShorthandProps<Props = {}, Ref = unknown, As extends keyof JSX.IntrinsicElements = never> = Props &
+  React.RefAttributes<Ref> & {
+    as?: As;
+    children?: Props['children'] | ShorthandRenderFunction<Props>;
+  };
+```
+
+## Implementation Details
+
+Changes are located mostly on `react-utilities/compose` methods and _Typings_.
+
+`ObjectShorthandProps` will have to support `as` and `ref` properties with Generics:
+
+```ts
+type ObjectShorthandProps<Props = {}, Ref = unknown, As extends keyof JSX.IntrinsicElements = never> = Props &
+  React.RefAttributes<Ref> & {
+    as?: As;
+    children?: Props['children'] | ShorthandRenderFunction<Props>;
+  };
+```
+
+All other _Typings_ would be adapted for the changes provided by `ObjectShorthandProps`.
+
+`getSlots` method would stop having special cases for `root` and only a simple `loop` around all provided `slots` would be enough.
+
+```ts
+export function getSlots<R extends ObjectShorthandPropsRecord>(
+  state: ComponentState<R>,
+  slotNames: (keyof R)[] = [],
+): {
+  slots: Slots<R>;
+  slotProps: SlotProps<R>;
+} {
+  const slots = {} as Slots<R>;
+  const slotProps = {} as SlotProps<R>;
+
+  for (const slotName of slotNames) {
+    const [slot, props] = getSlot(state, slotName);
+    slots[slotName] = slot;
+    slotProps[slotName] = props;
+  }
+  return { slots, slotProps: slotProps };
+}
+```
+
+## Pros and Cons
+
+### Pros
+
+1. Treating `root` as a simple slot simplifies _Typings_
+2. Simplification of _Typings_ also comes with better inference performance
+3. References available on every _slot_
+4. Not using `getNativeElementProps` means possible performance improvement
+
+### Cons
+
+1. [Widening Types problem](#⚠️-Widening-types-problem-⚠️)
+2. Explicit of _slots_ declaration makes easier for developers to make mistakes
```

Comment:
could you give some examples of these mistakes ?

Reply:
I can only think of two possible common mistakes right now:

- Since `root` becomes a regular slot, it might be possible developers forget to include it in the list of shorthands
- Since typings become a responsibility of the developer, typing errors might be prone, e.g: bad usage of `Partial` or `Required`

I'll add them